include/dali/core/mm/memory.h

```diff
@@ -0,0 +1,153 @@
+ */
+template <typename T, memory_kind kind, typename Context>
+mr_unique_ptr<T, kind, Context> alloc_raw(memory_resource<kind, Context> *mr, size_t count) {
+  size_t bytes = sizeof(T) * count;
+  size_t alignment = alignof(T);
+  void *mem = mr->allocate(bytes, alignment);
+  return mr_unique_ptr<T, kind, Context>(static_cast<T*>(mem),
+            mr_deallocator<kind, Context>{mr, bytes, alignment});
+}
+
+
+/**
+ * @brief Allocates uninitialized storage for `count` objects of type `T`
+ *
+ * This function allocates raw memory with suitable size and alignment to accommodate `count`
+ * object of type `T`. The memory is obtained from default memory resource.
+ * The size of memory is at least sizeof(T) * count and the alignment is at least alignof(T).
+ * The return value is a unique pointer with a deleter that will safely dispose of the allocated
+ * buffer, but does not call destructors of the objects.
+ *
+ * @param count     Number of objects for which the storage should suffice.
+ * @tparam T        Type of the object for which the storage is allocated.
+ * @tparam kind     The kind of requested memory.
+ * @tparam Context  The execution context in which the memory will be available.
+ */
+template <typename T, memory_kind kind>
+auto alloc_raw(size_t count) {
+  return alloc_raw<T>(GetDefaultResource<kind>(), count);
+}
+
+/**
+ * @brief Allocates uninitialized storage for `count` objects of type `T`
+ *        with stream-ordered semantics
+ *
+ * This function allocates memory with stream-ordered semantics. The `alloc_stream` denotes
+ * the stream on which the memory can be safely used without additional synchronization.
+ * Use the value `host_sync` if the memory needs to be accessible on all streams or on host as
+ * soon as the function returns.
+ * `dealloc_stream` denotest the stream for deallocation. Stream-ordered semantics guarantee,
+ * that if there's still some work pending on `dealloc_stream`, it will finish before the memory
+ * returned by this function is freed. Use `host_sync` for host-synchronous execution.
+ *
+ * This function allocates raw memory with suitable size and alignment to accommodate `count`
+ * object of type `T`. The memory is obtained from memory resource `mr` with stream semantics.
+ * The size of memory is at least sizeof(T) * count and the alignment is at least alignof(T).
+ * The return value is a unique pointer with a deleter that will safely dispose of the allocated
+ * buffer, but does not call destructors of the objects.
+ *
+ * @param mr              Memory resources to allocate the memory from
+ * @param count           Number of objects for which the storage should suffice
+ * @param alloc_stream    The CUDA stream on which the memory is immediately usable
+ * @param dealloc_stream  The CUDA stream which is guaranteed to finish all work scheduled
+ *                        before the deallocation of the memory.
```

Comment:
What would be a use case for different streams for allocation and deallocation?

Reply:
Well, the most common use case would be to allocate in host order and release in stream order (for CPU->GPU staging of pinned memory) and the reverse for D2H copy.
Similarly, if the data is filled on one stream and consumed on another, it should be released on that latter stream.
```C++
auto ptr = alloc_raw_async_unique<T>(mr, n, src_stream, dst_stream);
populate<<<grid, block, ..., src_stream>>>(ptr.get(), ...)
cudaEventRecord(event, src_stream);
cudaStreamWaitEvent(dst_stream, event);
consume<<grid, block, ..., dst_stream>>>(ptr.get(), ...);
// the pointer goes out of scope now and the deallocation is scheduled on dst_stream
```

Reply:
Understood